### DISCUSSION TITLE
Looking for memory consumption

### DISCUSSION THREAD

**Discussion by kornicameister** [simdjson/simdjson#1522] [Q&A]
Hey everyone,

I couldn't exactly find that in documentation or via benchmarks but I am interested to learn how simdjson actually deals with memory. I am pretty sure nothing is loaded by default right away in favor of lazy/dynamic loading. Question is though how memory consumption/management looks like if we have a 10GB array and we want to send out (like over HTTP) a chunks of 1000 elements, like in this Python-like pseudocode:

```
chunk = sim_array[start:stop]
send_over_http(chunk)
start = stop + 1
stop = stop + 1000
```

How many of items is, at once, in memory?

I hope, this is not a tedious question + I am sorry if that is written somewhere and I just managed my best to see it.

**Comment by lemire** [comment]
@kornicameister So that the question is clear, can you write your code using the actual `simdjson` C++ API? Our performance notes cover memory issues.

**Reply by kornicameister** [reply]
@lemire I am not actually using simdjson directly. I am using `pysimdjson` thatt @TkTech wrote.
Anyway, I did not want to suggest that you are having memory issues or anything like that.
Just wanted to get an understanding how much of a memory is allocated at a time.
Did not ask that question over at `pysimdjson` because, as far as I understand, is simply binds to C++ API and there's little of Python in there.

**Reply by lemire** [reply]
@kornicameister 

>  is simply binds to C++ API and there's little of Python in there.

The last time I checked, pysimdjson was, by default, mapping the JSON document into Python objects, strings and numbers... And this is likely to dominate your memory usage.

Even so, I am not sure that the code you describe could be done with simdjson. It could be, but I can't tell.


Here is a more general comment...

You can't reason about memory well unless you are specific.

It easy to measure the time something takes, but the memory usage is far trickier.

For example, you have to be aware of the distinction between virtual and real memory use. In C, I can write ...

```C++
char * b = malloc(100000000);
```

How much memory do you think that it uses? On many systems, the answer is zero. I mean, you are going to use 100000000 bytes of virtual memory at least, but most systems have 2^56 bytes of virtual memory (or just about) irrespective of the hardware (in other words, a nearly infinite amount).

So you can't look at the code and infer the memory usage, you have to count the allocated physical pages. And so then you have to be aware that memory is consumed in units of pages. So how much memory does this program take...

```C++
int main() {
  char * b = malloc(100);
  for(size_t i = 0; i < 100; i++) {
    b[i] = 1;
  }
  free(b);
}
```

The answer probably ranges to between 4kB to 1GB depending on your configured page size. And the page size varies quite a bit. Many ARM systems (phone, tablets) have 2MB pages, whereas conventional PCs often default on 4kB pages.

Thankfully, most operating systems provide you with tools that allow you to track the pages allocated to a program. That's great because you can reason specifically about the use case you care about.

Alternatively, you can provide C++ with your own memory allocator which you can use to study how it allocates memory. This might be tricky when used in conjunction with Python but if you have specific technical needs, that would be quite feasible.


Though Python is quite well made,  interpreters are generally memory hungry compared to C++ code. Suppose I start the Python interpreter, but I do nothing else. How much memory does that use? On my system the answer is 5.5MB of physical (real) memory for Python 3.6 and 4 MB for Python 2.7. Some of that memory usage can be shared though, so it probably a pessimistic take.

**Reply by kornicameister** [reply]
Really appreciate detailed answer. I am not C++ programmer so your answer here is actually really eye-opening. Thx a lot!

**Comment by TkTech** [comment]
Allowing end users of pysimdjson to answer this question is why I asked for https://github.com/simdjson/simdjson/issues/1017, which we'll hopefully get before 1.0.0. If I can plug in PyMem_Malloc, Python's own object/memory reporting tools would work for tracking simdjson.

> The last time I checked, pysimdjson was, by default, mapping the JSON document into Python objects, strings and numbers... And this is likely to dominate your memory usage.

The `load()` and `loads()` shortcuts do, as they just copy the built-in `json.load()`. None of the 'native' pysimdjson APIs do however, they return proxies.

That said, the overhead would be around 3x your document size if your input document is a `str`ing. In C, I encode the string you pass as UTF-8 and pass it into simdjson (which only understands UTF-8). CPython may keep your string in several different encodings internally, and there's no way to demand it use any one in particular. I then use simdjson's `padded_string`, which makes another copy (but this is hopefully gone soon with #1511, although only for the iteration API). The intermediate string is discarded at this point, and your original string in Python might be discarded as well if you're no longer using it.

**Reply by lemire** [reply]
> why I asked for #1017, which we'll hopefully get before 1.0.0

What one needs to do is to track the number of physical pages that have been allocated to the process. This is likely going to be an OS-specific call. It is possible that Python might have an API for that, but one needs to be aware that tracking virtual memory is useless.

**Reply by TkTech** [reply]
> tracking virtual memory is useless

I don't quite agree with that. Sure, how the system decides to allocate pages is no business of simdjson, but even if a page is merely reserved and not committed, it's still unavailable to the rest of the system.

Either way, python's built in `tracemalloc` or the `memory_profile` package give exact usage, and we can tag each allocation with a descriptor and domain of what it's being allocated for. PyMem_Malloc (python heap) and PyMem_RawMalloc (VirtualAlloc/malloc) are doing their own tracking.

**Reply by lemire** [reply]
>  it's still unavailable to the rest of the system.

On a legacy 32-bit system with a limited virtual space, or an embedded system without virtual memory, it might indeed be a concern.

**Reply by Artur-Biniek** [reply]
Hi lemire, does simdjson need to load entire file into RAM (at once) before it can parse it? Our use case is to translate big json files (think 30GB+) but we would like to keep working memory set of a process capped say at 1GB - is "streaming" possible with simdjson?

**Reply by lemire** [reply]
@Artur-Biniek If it is one big JSON document, then we do not support such a use case currently. I do not expect we will ever support it with our DOM API. We might support it with the new On Demand front-end at some point but it would assuredly have limitations since in the worst case, it requires mapping gigabytes of data in memory. From an engineering perspective, I would recommend you avoid monolithic multi-gigabyte JSON files.

If it is a stream of JSON documents (e.g., ndjson), then you can memory-map the input and stream through it using `parse_many`.
